File: rig/sparkplug-manager/src/catalog.rs
use std::collections::{BTreeMap, HashMap};

use anyhow::{Result, bail};

use txing_capability_protocol::InventoryDevice;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TypeCatalogDevice {
    pub thing_type: String,
    pub capabilities: Vec<String>,
    pub redcon_command_levels: Vec<u8>,
    pub redcon_rules: BTreeMap<u8, Vec<String>>,
}

impl TypeCatalogDevice {
    pub fn to_inventory_device(&self, thing_name: impl Into<String>) -> InventoryDevice {
        self.to_inventory_device_with_capabilities(thing_name, self.capabilities.clone())
    }

    pub fn to_inventory_device_with_capabilities(
        &self,
        thing_name: impl Into<String>,
        capabilities: Vec<String>,
    ) -> InventoryDevice {
        InventoryDevice {
            thing_name: thing_name.into(),
            thing_type: self.thing_type.clone(),
            capabilities,
            redcon_command_levels: self.redcon_command_levels.clone(),
            redcon_rules: self.redcon_rules.clone(),
        }
    }
}

pub fn reconstruct_type_record(parameters: &[(String, String)]) -> Result<TypeCatalogDevice> {
    let mut thing_type = None;
    let mut capabilities = None;
    let mut redcon_command_levels = None;
    let mut redcon_rules: BTreeMap<u8, Vec<String>> = BTreeMap::new();
    for (name, value) in parameters {
        let Some(leaf) = name.rsplit('/').next() else {
            continue;
        };
        match leaf {
            "thingType" => thing_type = Some(value.clone()),
            "capabilities" => capabilities = Some(parse_string_list(value)?),
            "redconCommandLevels" => {
                redcon_command_levels = Some(parse_redcon_levels(value)?);
            }
            _ => {
                if let Some(level_text) = name.split("/redconRules/").nth(1) {
                    let level_leaf = level_text
                        .split('/')
                        .next()
                        .ok_or_else(|| anyhow::anyhow!("redconRules leaf is malformed"))?;
                    let level = parse_redcon_level(level_leaf)?;
                    redcon_rules.insert(level, parse_string_list(value)?);
                }
            }
        }
    }
    let record = TypeCatalogDevice {
        thing_type: thing_type.ok_or_else(|| anyhow::anyhow!("missing thingType"))?,
        capabilities: capabilities.ok_or_else(|| anyhow::anyhow!("missing capabilities"))?,
        redcon_command_levels: redcon_command_levels
            .ok_or_else(|| anyhow::anyhow!("missing redconCommandLevels"))?,
        redcon_rules,
    };
    validate_type_record(&record)?;
    Ok(record)
}

pub fn validate_type_record(record: &TypeCatalogDevice) -> Result<()> {
    if record.thing_type.trim().is_empty() {
        bail!("thingType must not be empty");
    }
    if !record
        .capabilities
        .iter()
        .any(|capability| capability == "sparkplug")
    {
        bail!("capabilities must include sparkplug");
    }
    let capability_set: HashMap<&str, ()> = record
        .capabilities
        .iter()
        .map(|capability| (capability.as_str(), ()))
        .collect();
    for level in &record.redcon_command_levels {
        validate_redcon(*level)?;
    }
    for (level, capabilities) in &record.redcon_rules {
        validate_redcon(*level)?;
        if capabilities.is_empty() {
            bail!("redconRules.{level} must not be empty");
        }
        for capability in capabilities {
            if !capability_set.contains_key(capability.as_str()) {
                bail!("redconRules.{level} references unknown capability {capability}");
            }
        }
    }
    Ok(())
}

pub fn parse_string_list(value: &str) -> Result<Vec<String>> {
    if value.trim().is_empty() {
        return Ok(Vec::new());
    }
    value
        .split(',')
        .map(|item| {
            let text = item.trim();
            if text.is_empty() {
                bail!("list leaf contains an empty item");
            }
            Ok(text.to_string())
        })
        .collect()
}

fn parse_redcon_levels(value: &str) -> Result<Vec<u8>> {
    parse_string_list(value)?
        .into_iter()
        .map(|item| parse_redcon_level(&item))
        .collect()
}

fn parse_redcon_level(value: &str) -> Result<u8> {
    let level: u8 = value.parse()?;
    validate_redcon(level)?;
    Ok(level)
}

fn validate_redcon(level: u8) -> Result<()> {
    if !(1..=4).contains(&level) {
        bail!("redcon must be between 1 and 4");
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn reconstructs_redcon_rules_from_ssm_leaf_parameters() {
        let record = reconstruct_type_record(&[
            (
                "/txing/town/raspi/power/thingType".to_string(),
                "power".to_string(),
            ),
            (
                "/txing/town/raspi/power/capabilities".to_string(),
                "sparkplug,ble,power".to_string(),
            ),
            (
                "/txing/town/raspi/power/redconCommandLevels".to_string(),
                "4,3".to_string(),
            ),
            (
                "/txing/town/raspi/power/redconRules/4".to_string(),
                "sparkplug,ble".to_string(),
            ),
            (
                "/txing/town/raspi/power/redconRules/3".to_string(),
                "sparkplug,ble,power".to_string(),
            ),
        ])
        .unwrap();

        assert_eq!(record.thing_type, "power");
        assert_eq!(record.capabilities, vec!["sparkplug", "ble", "power"]);
        assert_eq!(record.redcon_command_levels, vec![4, 3]);
        assert_eq!(
            record.redcon_rules.get(&3).unwrap(),
            &vec![
                "sparkplug".to_string(),
                "ble".to_string(),
                "power".to_string()
            ]
        );
    }
}

File: rig/internal/catalog/catalog.go
package catalog

import (
	"fmt"
	"strconv"
	"strings"

	"github.com/mparkachov/txing/rig/internal/protocol"
)

type TypeCatalogDevice struct {
	ThingType           string
	Capabilities        []string
	RedconCommandLevels []uint8
	RedconRules         map[uint8][]string
}

func (d TypeCatalogDevice) ToInventoryDevice(thingName string) protocol.InventoryDevice {
	return d.ToInventoryDeviceWithCapabilities(thingName, append([]string(nil), d.Capabilities...))
}

func (d TypeCatalogDevice) ToInventoryDeviceWithCapabilities(thingName string, capabilities []string) protocol.InventoryDevice {
	return protocol.InventoryDevice{
		ThingName:           thingName,
		ThingType:           d.ThingType,
		Capabilities:        capabilities,
		RedconCommandLevels: append([]uint8(nil), d.RedconCommandLevels...),
		RedconRules:         cloneRules(d.RedconRules),
	}
}

func ReconstructTypeRecord(parameters [][2]string) (TypeCatalogDevice, error) {
	var record TypeCatalogDevice
	record.RedconRules = map[uint8][]string{}
	for _, parameter := range parameters {
		name := parameter[0]
		value := parameter[1]
		parts := strings.Split(name, "/")
		leaf := ""
		if len(parts) > 0 {
			leaf = parts[len(parts)-1]
		}
		switch leaf {
		case "thingType":
			record.ThingType = value
		case "capabilities":
			items, err := ParseStringList(value)
			if err != nil {
				return TypeCatalogDevice{}, err
			}
			record.Capabilities = items
		case "redconCommandLevels":
			levels, err := parseRedconLevels(value)
			if err != nil {
				return TypeCatalogDevice{}, err
			}
			record.RedconCommandLevels = levels
		default:
			before, after, ok := strings.Cut(name, "/redconRules/")
			_ = before
			if !ok {
				continue
			}
			levelText := strings.Split(after, "/")[0]
			level, err := parseRedconLevel(levelText)
			if err != nil {
				return TypeCatalogDevice{}, err
			}
			items, err := ParseStringList(value)
			if err != nil {
				return TypeCatalogDevice{}, err
			}
			record.RedconRules[level] = items
		}
	}
	if err := ValidateTypeRecord(record); err != nil {
		return TypeCatalogDevice{}, err
	}
	return record, nil
}

func ValidateTypeRecord(record TypeCatalogDevice) error {
	if strings.TrimSpace(record.ThingType) == "" {
		return fmt.Errorf("thingType must not be empty")
	}
	if !contains(record.Capabilities, "sparkplug") {
		return fmt.Errorf("capabilities must include sparkplug")
	}
	capabilitySet := map[string]struct{}{}
	for _, capability := range record.Capabilities {
		capabilitySet[capability] = struct{}{}
	}
	for _, level := range record.RedconCommandLevels {
		if err := protocol.ValidateRedcon(level, "redconCommandLevels"); err != nil {
			return err
		}
	}
	for level, capabilities := range record.RedconRules {
		if err := protocol.ValidateRedcon(level, "redconRules"); err != nil {
			return err
		}
		if len(capabilities) == 0 {
			return fmt.Errorf("redconRules.%d must not be empty", level)
		}
		for _, capability := range capabilities {
			if _, ok := capabilitySet[capability]; !ok {
				return fmt.Errorf("redconRules.%d references unknown capability %s", level, capability)
			}
		}
	}
	return nil
}

func ParseStringList(value string) ([]string, error) {
	if strings.TrimSpace(value) == "" {
		return []string{}, nil
	}
	raw := strings.Split(value, ",")
	items := make([]string, 0, len(raw))
	for _, item := range raw {
		text := strings.TrimSpace(item)
		if text == "" {
			return nil, fmt.Errorf("list leaf contains an empty item")
		}
		items = append(items, text)
	}
	return items, nil
}

func parseRedconLevels(value string) ([]uint8, error) {
	items, err := ParseStringList(value)
	if err != nil {
		return nil, err
	}
	levels := make([]uint8, 0, len(items))
	for _, item := range items {
		level, err := parseRedconLevel(item)
		if err != nil {
			return nil, err
		}
		levels = append(levels, level)
	}
	return levels, nil
}

func parseRedconLevel(value string) (uint8, error) {
	parsed, err := strconv.ParseUint(value, 10, 8)
	if err != nil {
		return 0, err
	}
	level := uint8(parsed)
	if err := protocol.ValidateRedcon(level, "redcon"); err != nil {
		return 0, err
	}
	return level, nil
}

func contains(values []string, target string) bool {
	for _, value := range values {
		if value == target {
			return true
		}
	}
	return false
}

func cloneRules(input map[uint8][]string) map[uint8][]string {
	output := make(map[uint8][]string, len(input))
	for key, value := range input {
		output[key] = append([]string(nil), value...)
	}
	return output
}
